BUG: pivot_table strings as aggfunc

bobhaffner opened the pull request:
- [x] closes #18713
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

bobhaffner committed:
pivot_table strings as aggfunc

pep8speaks commented:
Hello @bobhaffner! Thanks for submitting the PR.

 - In the file [`pandas/core/reshape/pivot.py`](https://github.com/pandas-dev/pandas/blob/8b81dabb2651133d0602773f8d9ab825d27ee513/pandas/core/reshape/pivot.py), following are the PEP8 issues :

> [Line 42:13](https://github.com/pandas-dev/pandas/blob/8b81dabb2651133d0602773f8d9ab825d27ee513/pandas/core/reshape/pivot.py#L42): [E265](https://duckduckgo.com/?q=pep8%20E265) block comment should start with '# '

bobhaffner committed:
fixed pep8 issues

gfyoung commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -321,7 +321,8 @@ Reshaping
 - Fixed construction of a :class:`Series` from a ``dict`` containing ``NaN`` as key (:issue:`18480`)
 - Bug in :func:`Series.rank` where ``Series`` containing ``NaT`` modifies the ``Series`` inplace (:issue:`18521`)
--
+- Bug in :func:`Dataframe.pivot_table` which fails when you pass a string for aggfunc arg (:issue:`18713`)
```
We generally don't use the second-person (or any such person-pronouns in general).

bobhaffner committed:
fixed whatsnew

bobhaffner commented:
Thanks for the feedback @gfyoung.

gfyoung commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -1109,6 +1109,20 @@ def test_pivot_margins_name_unicode(self):
+        stds = f(np.std)
+        expected = concat([means, stds], keys=['mean', 'std'], axis=1)
+        tm.assert_frame_equal(result, expected)
+
```
I like what you're testing here, but I would modify a couple of things:

1) Add newlines after each `tm.assert_frame_equal`.  It makes it easier to read.
2) Construct the output explicitly using the `DataFrame` constructor (and check against that too).  At this point, I know for example that `f('sum')` and `f(np.sum)` should be the same, but I don't know what the actual result is.  e.g.

~~~python
result = f('sum')
func_result = f(np.sum)
expected = DataFrame(...)
tm.assert_frame_equal(result, func_result)  # consistency
tm.assert_frame_equal(result, expected)  # correctness
~~~

jreback replied:
you can parametrize this test by passing these functions in as well

something like

```
@pytest.mark.parametrize("f, f_numpy", [('sum', np.sum), ('mean', np.mean'), ('std', lambda x: np.std(x, ddof=0))])
def test_pivot_func_string(self, f, f_numpy):
```

note that ``std`` will resolve to pandas ``std`` and so ``np.std`` needs handling here (for the test)

bobhaffner replied:
Thank you, both,  I'll incorporate these changes

bobhaffner committed:
updated tests

bobhaffner commented:
`test_pivot_string_as_func` is an attempt at @gfyoung's suggestion regarding comparing to a frame created from `DataFrame()` 

`test_pivot_string_func_vs_func` is an attempt at @jreback's suggestion of using `@pytest.mark.parametrize` to compare string_funcs to numpy funcs

Separating the cases into two functions seemed like the right thing to do.  As opposed to one rather long one with multiple `DataFrame` statements

gfyoung commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -1109,6 +1109,48 @@ def test_pivot_margins_name_unicode(self):
+             (['sum', 'mean'], [np.sum, np.mean]),
+             (['sum', 'std'], [np.sum, np.std]),
+             (['std', 'mean'], [np.std, np.mean])]
+
```
Remove newline.

bobhaffner replied:
This is a weird one, flake8 likes that newline

jreback replied:
instead of using funcs, simply put this in the parametrize

gfyoung commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -1109,6 +1109,48 @@ def test_pivot_margins_name_unicode(self):
+
+    @pytest.mark.parametrize("f, f_numpy", funcs)
+    def test_pivot_string_func_vs_func(self, f, f_numpy):
+        # GH #18713
```
Add that this test is for consistency purposes.

gfyoung commented on pandas/tests/reshape/test_pivot.py:
```diff
@@ -1109,6 +1109,48 @@ def test_pivot_margins_name_unicode(self):
         tm.assert_frame_equal(table, expected)
 
+    def test_pivot_string_as_func(self):
+        # GH #18713
```
Add that this test is for correctness.

gfyoung commented:
> Separating the cases into two functions seemed like the right thing to do. As opposed to one rather long one with multiple DataFrame statements.

Seems reasonable to me.

bobhaffner committed:
added comments to test funcs

bobhaffner commented:
Just following up on this to make sure no other changes are required.  No worries if folks are busy

bobhaffner committed:
removed list usage in test parametrize

jreback commented on doc/source/whatsnew/v0.22.0.txt:
```diff
@@ -341,7 +341,8 @@ Reshaping
 - Fixed construction of a :class:`Series` from a ``dict`` containing ``NaN`` as key (:issue:`18480`)
 - Bug in :func:`Series.rank` where ``Series`` containing ``NaT`` modifies the ``Series`` inplace (:issue:`18521`)
--
+- Bug in :func:`Dataframe.pivot_table` which fails when the ``aggfunc`` arg is of type string.  The behavior is now consistent with other methods like ``agg`` and ``apply`` (:issue:`18713`)
```
rebase on master. can you move to 0.23 (docs were renamed), prob easiest to just check this file from master and past in new one

bobhaffner commented:
Very interesting.  Looks like the rebase handled the whatsnew change quite nicely

```
(pandas-dev) Bobs-MacBook-Pro:pandas-dev bob$ git rebase upstream/master
First, rewinding head to replay your work on top of it...
Applying: pivot_table strings as aggfunc
Using index info to reconstruct a base tree...
A	doc/source/whatsnew/v0.22.0.txt
.git/rebase-apply/patch:11: trailing whitespace.
- 
warning: 1 line adds whitespace errors.
Falling back to patching base and 3-way merge...
Auto-merging doc/source/whatsnew/v0.23.0
Applying: fixed pep8 issues
Applying: fixed whatsnew
Using index info to reconstruct a base tree...
A	doc/source/whatsnew/v0.22.0.txt
Falling back to patching base and 3-way merge...
Auto-merging doc/source/whatsnew/v0.23.0
Applying: updated tests
Applying: added comments to test funcs
Applying: removed list usage in test parametrize
```

gfyoung commented:
@bobhaffner : This is `git` - you should expect great things from it :smile: (unless you're pushing large files OR operating at big-tech-company-scale... :0 )

bobhaffner commented:
@gfyoung Haha, gotcha  :-)

jreback commented:
lgtm. ping on green.

bobhaffner commented:
> lgtm. ping on green.

@jreback is this intended for me or some kind of shorthand for a notification process?   At any rate, ping   :-)    

Looks like there's a conflict with the whatsnew piece being changed back to the old name.  I can follow suit or wait till this is sorted out.

jreback committed:
Merge branch 'master' into PR_TOOL_MERGE_PR_18810

jreback committed:
doc edit

jreback commented:
thanks @bobhaffner 

yes 'ping on green' is a notification of you to ping me when this has passed and all buttons are green.

bobhaffner commented:
Thanks @jreback and @gfyoung!!!
